lib/jwt/jwt.go

```diff
@@ -0,0 +1,284 @@
+		Config:    config,
+		publicKey: publicKey,
+	}, nil
+}
+
+// SignParams are the claims to be embedded within the JWT token.
+type SignParams struct {
+	// Username is the Teleport identity.
+	Username string
+
+	// Roles are the roles assigned to the user within Teleport.
+	Roles []string
+
+	// Expiry is time to live for the token.
+	Expiry time.Duration
+}
+
+// Check verifies all the values are valid.
+func (p *SignParams) Check() error {
+	if p.Username == "" {
+		return trace.BadParameter("missing username")
+	}
+	if len(p.Roles) == 0 {
+		return trace.BadParameter("missing roles")
+	}
+	if p.Expiry == 0 {
+		return trace.BadParameter("expiry required")
+	}
+
+	return nil
+}
+
+// Sign will return a signed JSON token with the passed in claims embedded
+// within the token.
+func (k *Key) Sign(p *SignParams) (string, error) {
+	if k.privateKey == nil {
+		return "", trace.BadParameter("can not sign token with non-signing key")
+	}
+	if err := p.Check(); err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	signingKey := jose.SigningKey{
+		Algorithm: jose.RS256,
+		Key:       k.privateKey,
+	}
+
+	sig, err := jose.NewSigner(signingKey, (&jose.SignerOptions{}).WithType("JWT"))
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	claims := jwtClaims{
+		Claims: josejwt.Claims{
+			Subject:   p.Username,
+			Issuer:    "Gravitational Teleport",
+			NotBefore: jwt.NewNumericDate(k.Clock.Now().Add(-10 * time.Second)),
+			Expiry:    jwt.NewNumericDate(k.Clock.Now().Add(p.Expiry)),
+		},
+		Username: p.Username,
+		Roles:    p.Roles,
+	}
+	token, err := josejwt.Signed(sig).Claims(claims).CompactSerialize()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return token, nil
+}
+
+// Verify will validate the passed in JWT token.
+func (k *Key) Verify(raw string) (*jwtClaims, error) {
+	if k.publicKey == nil {
+		return nil, trace.BadParameter("can not verify token without public key")
+	}
+
+	tok, err := josejwt.ParseSigned(raw)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var out jwtClaims
+	if err := tok.Claims(k.publicKey, &out); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	expectedClaims := jwt.Expected{
+		Time: k.Clock.Now(),
+	}
+	if err = out.Validate(expectedClaims); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &out, nil
+}
+
+// Claims defines the properties that can be extracted from a JWT token.
+type Claims interface {
+	// GetUsername returns the Teleport identity of the user.
+	GetUsername() string
+
+	// GetRoles returns the list of roles assigned to the user within Teleport.
+	GetRoles() []string
+}
+
+type jwtClaims struct {
+	josejwt.Claims
+
+	Username string   `json:"username"`
+	Roles    []string `json:"roles"`
+}
+
+// GetUsername returns the Teleport identity of the user.
+func (c *jwtClaims) GetUsername() string {
+	return c.Username
+}
+
+// GetRoles returns the list of roles assigned to the user within Teleport.
+func (c *jwtClaims) GetRoles() []string {
+	return c.Roles
+}
```

Comment:
how about making `jwtClaims` the exported type (named `Claims`) and removing the interface?
it's a simple-enough struct to fake, easier than making a mock interface implementation